File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency/Persistence/src/main/java/impl/database/PurchaseTable.java
package impl.database;

import api.Constants;
import api.Table;
import domain.dto.PurchaseDto;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class PurchaseTable implements Table<Long, PurchaseDto> {

    private final Connection connection;
    private Map<Constants.Db.Queries, PreparedStatement> statements;
    private Map<Constants.Db.Queries, PreparedStatement> statementsTourist;
    private static final Logger logger = LogManager.getLogger();

    public PurchaseTable(Connection connection) throws SQLException {
        logger.info("Initializing PurchaseTable");
        this.statements = new HashMap<>();
        this.statementsTourist = new HashMap<>();
        this.connection = connection;
        initStatements();
    }

    void initStatements() throws SQLException {
        statements.put(Constants.Db.Queries.ADD,
                connection.prepareStatement("INSERT INTO purchase (id, flightId, clientName, clientAddress, nrOfSeats) VALUES (?, ?, ?, ?, ?)"
                ));
        statements.put(Constants.Db.Queries.DELETE,
                connection.prepareStatement("DELETE FROM purchase WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.FIND_BY_ID,
                connection.prepareStatement("SELECT * FROM purchase WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.GET_ALL,
                connection.prepareStatement("SELECT * FROM purchase"
                ));

        statementsTourist.put(Constants.Db.Queries.ADD,
                connection.prepareStatement("INSERT INTO tourist (name, purchaseId) VALUES (?, ?)"
                ));
        statementsTourist.put(Constants.Db.Queries.FIND_BY_ID,
                connection.prepareStatement("SELECT * FROM tourist WHERE purchaseId = ?"
                ));

    }
    @Override
    public void add(PurchaseDto purchaseDto) {
        logger.traceEntry("saving purchase {} ", purchaseDto);
        PreparedStatement statement = statements.get(Constants.Db.Queries.ADD);
        try {
            statement.setLong(1, purchaseDto.getId());
            statement.setLong(2, purchaseDto.getFlightId());
            statement.setString(3, purchaseDto.getClientName());
            statement.setString(4, purchaseDto.getClientAddress());
            statement.setInt(5, purchaseDto.getNrOfSeats());
            addTourists(purchaseDto.getId(), purchaseDto.getTourists());
            int result = statement.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    private void addTourists(Long id, List<String> tourists) {
        logger.traceEntry("saving tourist {} ", tourists);
        for (String tourist: tourists) {
            PreparedStatement statement = statementsTourist.get(Constants.Db.Queries.ADD);
            try {
                statement.setString(1, tourist);
                statement.setLong(2, id);
                int result = statement.executeUpdate();
                logger.trace("Saved {} instances", result);
            } catch (SQLException e) {
                logger.error(e);
                System.err.println("Error DB " + e);
            }
        }
        logger.traceExit();
    }

    @Override
    public void delete(Long o) {
        //TODO
    }

    @Override
    public void update(PurchaseDto elem, PurchaseDto newElem) {

    }

    @Override
    public Optional<PurchaseDto> findById(Long id) {
        logger.traceEntry();
        List<PurchaseDto> purchaseDtos = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.FIND_BY_ID);
        try {
            statement.setLong(1, id);
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    long flightId = result.getLong("flightId");
                    String clientName = result.getString("clientName");
                    String clientAddress = result.getString("clientAddress");
                    int nrOfSeats = result.getInt("nrOfSeats");
                    List<String> tourists = getTourists(id);

                    PurchaseDto purchaseDto = new PurchaseDto(flightId, clientName, clientAddress, tourists, nrOfSeats);
                    purchaseDto.setId(id1);
                    purchaseDtos.add(purchaseDto);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return Optional.of(purchaseDtos.get(0));
    }

    private List<String> getTourists(Long id) {
        logger.traceEntry();
        List<String> tourists = new ArrayList<>();
        PreparedStatement statement = statementsTourist.get(Constants.Db.Queries.FIND_BY_ID);
        try {
            statement.setLong(1, id);
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    String name = result.getString("name");
                    tourists.add(name);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return tourists;
    }

    @Override
    public List<PurchaseDto> getAll() {
        logger.traceEntry();
        List<PurchaseDto> purchaseDtos = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.GET_ALL);
        try {
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    long flightId = result.getLong("flightId");
                    String clientName = result.getString("clientName");
                    String clientAddress = result.getString("clientAddress");
                    int nrOfSeats = result.getInt("nrOfSeats");
                    List<String> tourists = getTourists(id1);

                    PurchaseDto purchaseDto = new PurchaseDto(flightId, clientName, clientAddress, tourists, nrOfSeats);
                    purchaseDto.setId(id1);
                    purchaseDtos.add(purchaseDto);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return purchaseDtos;
    }
}

File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency2/persistance/database/PurchaseTable.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using model.Dto;
using model.Utils;
using travelAgency2.Repository.Database;

namespace persistance.Database
{
    internal class PurchaseTable : ITable<long, PurchaseDto>
    {
        private SQLiteConnection connection;
        private Dictionary<Db.Queries, SQLiteCommand> statements;
        private Dictionary<Db.Queries, SQLiteCommand> statementsTourist;
        private static readonly log4net.ILog logger = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        public PurchaseTable(SQLiteConnection connection)
        {
            logger.Info("Initializing PurchaseTable");
            this.connection = connection;
            this.statements = new Dictionary<Db.Queries, SQLiteCommand>();
            this.statementsTourist = new Dictionary<Db.Queries, SQLiteCommand>();
            InitStatements();
        }

        void InitStatements()
        {
            SQLiteCommand command1 = new SQLiteCommand(connection);
            SQLiteCommand command2 = new SQLiteCommand(connection);
            SQLiteCommand command3 = new SQLiteCommand(connection);
            SQLiteCommand command4 = new SQLiteCommand(connection);
            SQLiteCommand command5 = new SQLiteCommand(connection);
            SQLiteCommand command6 = new SQLiteCommand(connection);
            command1.CommandText = "INSERT INTO purchase (id, flightId, clientName, clientAddress, nrOfSeats) " +
                "VALUES (@id, @flightId, @clientName, @clientAddress, @nrOfSeats)";
            command2.CommandText = "DELETE FROM purchase WHERE id = @id";
            command3.CommandText = "SELECT FROM purchase WHERE id = @id";
            command4.CommandText = "SELECT * FROM purchase";
            command5.CommandText = "INSERT INTO tourist(name, purchaseId) VALUES(?, ?)";
            command6.CommandText = "SELECT * FROM tourist WHERE purchaseId = ?";
            statements.Add(Db.Queries.ADD, command1);
            statements.Add(Db.Queries.DELETE, command2);
            statements.Add(Db.Queries.FIND_BY_ID, command3);
            statements.Add(Db.Queries.GET_ALL, command4);
            statementsTourist.Add(Db.Queries.ADD, command5);
            statementsTourist.Add(Db.Queries.FIND_BY_ID, command6);
        }
        public void AddTourists(long id, List<String> tourists)
        {
            logger.Info("enter in add tourist");
            SQLiteCommand statement = statementsTourist[Db.Queries.ADD];
            foreach (String tour in tourists)
            {
                try
                {
                    statement.Parameters.AddWithValue("@name", tour);
                    statement.Parameters.AddWithValue("@purchaseId", id);
                    int numberofrowsaffected = statement.ExecuteNonQuery();
                    logger.InfoFormat("Saved {0} instances", numberofrowsaffected);
                }
                catch (Exception ex)
                {
                    logger.Error(ex);
                    Console.WriteLine(ex.Message);
                }

            }
            logger.Info("exit from add tourist");
        }
        public void Add(PurchaseDto purchaseDto)
        {
            logger.Info("enter in add purchase");
            SQLiteCommand statement = statements[Db.Queries.ADD];
            try
            {
                statement.Parameters.AddWithValue("@id", purchaseDto.Id);
                statement.Parameters.AddWithValue("@flightId", purchaseDto.flightId);
                statement.Parameters.AddWithValue("@clientName", purchaseDto.clientName);
                statement.Parameters.AddWithValue("@clientAddress", purchaseDto.clientAddress);
                statement.Parameters.AddWithValue("@nrOfSeats", purchaseDto.nrOfSeats);
                AddTourists(purchaseDto.Id, purchaseDto.tourists);
                int numberofrowsaffected = statement.ExecuteNonQuery();
                logger.InfoFormat("Saved {0} instances", numberofrowsaffected);
            }
            catch (Exception ex)
            {
                logger.Error(ex);
                Console.WriteLine(ex.Message);
            }
            logger.Info("exit from add purchase");
        }

        public void Delete(long id)
        {
            //TODO
        }
        public void Update(PurchaseDto purchaseDto, PurchaseDto purchaseDto1)
        {
            //TODO
        }
        public List<String> getTourists(long id)
        {
            logger.Info("enter in getTourists");
            List<String> tourists = new List<String>();
            SQLiteCommand statement = statementsTourist[Db.Queries.FIND_BY_ID];
            statement.Parameters.AddWithValue("@id", id);
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    tourists.Add(reader.GetString(0));
                }
            }
            reader.Close();
            logger.Info("exit from get tourists");
            return tourists;
        }
        public PurchaseDto FindById(long id)
        {
            logger.Info("enter in findById purchase");
            List<PurchaseDto> purchaseDtos = new List<PurchaseDto>();
            SQLiteCommand statement = statements[Db.Queries.FIND_BY_ID];
            statement.Parameters.AddWithValue("@id", id);
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    List<String> tourists = getTourists(id);
                    PurchaseDto purchaseDto = new PurchaseDto((long) reader.GetInt64(0),
                        (long)reader.GetInt64(1), reader.GetString(2), reader.GetString(3), tourists, reader.GetInt32(4));
                    purchaseDtos.Add(purchaseDto);
                }
            }
            reader.Close();
            logger.Info("exit from findById purchase");
            return purchaseDtos[0];
        }

        public List<PurchaseDto> GetAll()
        {
            logger.Info("enter in getAll purchase");
            List<PurchaseDto> purchaseDtos = new List<PurchaseDto>();
            SQLiteCommand statement = statements[Db.Queries.GET_ALL];
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    long id = (long)reader.GetInt64(0);
                    List<String> tourists = getTourists(id);
                    PurchaseDto purchaseDto = new PurchaseDto(id,
                        (long)reader.GetInt64(1), reader.GetString(2), reader.GetString(3), tourists, reader.GetInt32(4));
                    purchaseDtos.Add(purchaseDto);
                }
            }
            reader.Close();
            logger.Info("exit from getAll purchase");
            return purchaseDtos;
        }
    }
}
